libevmasm/Assembly.cpp

```diff
@@ -579,17 +579,49 @@ std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSO
 			{
 				shared_ptr<Assembly> subassembly(Assembly::fromJSON(code, sourceList, _level + 1).first);
 				solAssert(subassembly);
-				result->m_subs.emplace_back(make_shared<Assembly>(*subassembly));
-				// TODO: this shouldn't be enough for the general case.
-				result->encodeSubPath({0, 0});
+				result->m_subs.emplace_back(subassembly);
 			}
 			else
 				solThrow(AssemblyImportException, "Key inside '.data' '" + dataItemID + "' can only be a valid hex-string or an object.");
 		}
 	}
+	if (_level == 0)
+		result->updatePaths();
 	return std::make_pair(result, sourceList);
 }
 
+void Assembly::updatePaths(std::vector<Assembly*> const& _parents, std::vector<size_t> const& _absolutePathFromRoot)
```

Comment:
@ekpyron @cameel I implemented yesterday a more generic version to deal with the sub-paths to assembly objects that where imported via yul. It seem to work, but I would like to have a better test coverage for that (this is basically only used if evm json was exported from yul and the resulting evm json gets imported again - I did not manage to trigger this with plain solidity). However, because of that I would like to add at least one more complicated command-line test and I guess it cannot hurt to also extend the import/export script to also reimport all of our yul sources and check also those for byte-code equivalence.

Reply:
I just spent a while trying to understand what this does. If I understand correctly, it runs `encodeSubPath()` for all possible subpaths across the whole subassembly tree? Looks like this should work correctly, but it's a total overkill :)

From what I can tell, the subpaths are relevant only for assemblies referenced in `dataOffset()` and `dataSize()`. So it seems like the proper solution would be to handle it by calling `encodeSubPath()` while the corresponding `PushSub` and `PushSubSize` are processed in `createAssemblyItemFromJSON()`.

I guess we can leave it as is in the initial version, but this is something that we should really refactor in a separate PR.

@r0qs Looks like this is a good reason to actually have some coverage for assembly created with `--via-ir` tests. Doing that for all snippets from the repo is too much, but I think we should at least have a single CLI test that compiles some Solidity contract via IR and imports it. The contract should use `new` so that we get more subassemblies inside. I guess this could also be left for a separate PR as long as you test manually that this works.

Reply:
Actually looks like it is covered by some CLI tests already.

Also, it seems that what we ended up with here is just a quick and dirty solution to get this merged. I would still be good to refactor this though. Previous thread: https://github.com/ethereum/solidity/pull/13673#pullrequestreview-1388117871